Cubical/HITs/Ints/BiInvInt/Properties.agda

```diff
@@ -3,146 +3,240 @@ module Cubical.HITs.Ints.BiInvInt.Properties where
 
 open import Cubical.Core.Everything
 
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Equiv.Properties
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Isomorphism
 open import Cubical.Foundations.Prelude
 
-open import Cubical.Data.Nat hiding (_+_; +-comm)
-open import Cubical.Data.Int
+open import Cubical.Data.Nat using (ℕ)
+import Cubical.Data.Int as Int
 open import Cubical.Data.Bool
 
 open import Cubical.HITs.Ints.BiInvInt.Base
 
+infixl 6 _+_ _-_
+infixl 7 _*_
+
+-- To prove a property P about BiInvInt, we need to show:
+-- * P zero
+-- * If P n, then P (suc n)
+-- * If P n, then P (pred n)
+BiInvInt-ind-prop :
+  ∀ {ℓ} {P : BiInvInt → Type ℓ} → (∀ n → isProp (P n)) →
+  P zero → (∀ n → P n → P (suc n)) → (∀ n → P n → P (pred n)) →
+  (n : BiInvInt) → P n
+BiInvInt-ind-prop {P = P} P-isProp P-zero P-suc P-pred = φ
+  where
+  P-predr : ∀ n → P n → P (predr n)
+  P-predr n x = subst P (predl≡predr n) (P-pred n x)
+
+  P-predl : ∀ n → P n → P (predl n)
+  P-predl = P-pred
+
+  P-isProp' : {a b : BiInvInt} (p : a ≡ b) (x : P a) (y : P b) → PathP (λ i → P (p i)) x y
+  P-isProp' _ _ _ = toPathP (P-isProp _ _ _)
+
+  φ : (n : BiInvInt) → P n
+  φ zero = P-zero
+  φ (suc n) = P-suc n (φ n)
+  φ (predr n) = P-predr n (φ n)
+  φ (suc-predr n i) = P-isProp' (suc-predr n) (P-suc (predr n) (P-predr n (φ n))) (φ n) i
+  φ (predl n) = P-predl n (φ n)
+  φ (predl-suc n i) = P-isProp' (predl-suc n) (P-predl (suc n) (P-suc n (φ n))) (φ n) i
+
+-- To define a function BiInvInt → A, we need:
+-- * a point z : A for zero
+-- * an equivalence s : A ≃ A for suc/pred
+BiInvInt-rec : ∀ {ℓ} {A : Type ℓ} → A → A ≃ A → BiInvInt → A
+BiInvInt-rec {A = A} z e = φ
+  where
+  e-Iso : Iso A A
+  e-Iso = equivToIso e
+
+  φ : BiInvInt → A
+  φ zero = z
+  φ (suc n) = Iso.fun e-Iso (φ n)
+  φ (predr n) = Iso.inv e-Iso (φ n)
+  φ (suc-predr n i) = Iso.rightInv e-Iso (φ n) i
+  φ (predl n) = Iso.inv e-Iso (φ n)
+  φ (predl-suc n i) = Iso.leftInv e-Iso (φ n) i
+
+sucEquiv : BiInvInt ≃ BiInvInt
+sucEquiv = isoToEquiv (iso suc pred suc-predl predl-suc)
+
 -- addition
 
-_+ᴮ_ : BiInvInt → BiInvInt → BiInvInt
-m +ᴮ zero          = m
-m +ᴮ suc n         = suc (m +ᴮ n)
-m +ᴮ predr n       = predr (m +ᴮ n)
-m +ᴮ predl n       = predl (m +ᴮ n)
-m +ᴮ suc-predr n i = suc-predr (m +ᴮ n) i
-m +ᴮ predl-suc n i = predl-suc (m +ᴮ n) i
+-- the following equalities hold definitionally:
+--   zero   + n ≡ n
+--   suc m  + n ≡ suc (m + n)
+--   pred m + n ≡ pred (m + n)
+_+_ : BiInvInt → BiInvInt → BiInvInt
+_+_ = BiInvInt-rec (idfun BiInvInt) (postCompEquiv sucEquiv)
 
```

Comment:
Has it not been a problem that `m + n` would often get expanded to the `BiInvInt-rec ...` call?

Reply:
I'm not sure what you mean. How would this problem show up?

Reply:
If `m + n` shows up in a goal type, Agda might expand it, it will if you ask for the normal form but in other situations too.

Also, it could have shown up as slower typechecking times.

Reply:
Oh I see. It doesn't get expanded in (non-normalized) goal types, and the file typechecks in 3 seconds. But, when I try `C-u C-u C-c C-t` in the following goal, it takes over 30 seconds and spits out the following 66598-line term: https://gist.github.com/knrafto/ed60e415db86989c0200f9e6397ea125
(I'm not sure how much of that time is just printing though)

```
*-comm : ∀ m n → m * n ≡ n * m
*-comm = BiInvInt-ind-prop (λ _ → isPropΠ λ _ → isSetBiInvInt _ _)
  (λ n → sym (*-zero n))
  (λ m p n → {!!})  -- cong (n +_) (p n) ∙ sym (*-suc n m)
  (λ m p n → cong (- n +_) (p n) ∙ sym (*-pred n m))
```

I guess I never tried to normalize anything because the default goal types were pretty readable already

Reply:
Right, that's also made worse by the use of the `iso` constructor rather than doing a separate definition by copatterns.

Reply:
Can you then make separate copattern definitions for the isos needed in `sucEquiv` and `isEquiv-n+` ?
We are trying to stick to that as a general policy to keep normal forms smaller.

I forgot to thank you for the PR btw! It's a great use of abstraction, and hopefully any problems can be improved on by the tooling.

Reply:
Thank you! And thanks for your work on cubical agda, it's been fun to use.

I created separate definitions for the isos. I haven't really used copatterns before, but I was misled by [the docs](https://agda.readthedocs.io/en/v2.6.1/language/copatterns.html) into thinking copatterns are syntactic sugar for using the constructor. Why does using copatterns result in faster typechecking? Is that a bug?

Reply:
it's not a bug, the reason is that `sucIso` defined by copatterns is already in weak head normal form, so it only reduces when projected from.

Records having eta rules by default still means that the equality is unchanged, but Agda tries to only apply those when useful. 

One can go further with `no-eta-equality` but that's not been necessary yet.

Reply:
Sorry I mean, is it a bug that performance is so much worse for the constructor form? Maybe Agda should be less aggressive when expanding the definitions under the `iso`, for things like definitional equality and printing the normalized goal type.

Reply:
Oh, yeah, Agda should be more careful about expanding definitions in general, and/or there should be good expressive ways to control expansion. it would be a substantial project to improve the situation though.